добавил функция «Поиск»

rifnurV opened the pull request.

rifnurV committed:
добавил функция «Поиск»

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -117,4 +117,15 @@ public Set<Long> getFilmLikesByFilmId(Long filmId) {
+    public List<Film> searchFilms(String query, Boolean searchByTitle) {
+        List<Film> films = List.of();
+        if (searchByTitle){
+            films = jdbcTemplate.query("select * from film f where lower(f.name) like ? " +
```
можно через if(searchByTitle) else, чтобы не делать инициализацию
в else вернуть `Collections.emptyList();`

Также сделать query через `""" sql query """`

rifnurV replied:
добавил условие else

yusuf33nn commented:
Тесты + зелёная джоба по постман коллекции

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/service/DefaultFilmService.java:
```diff
@@ -95,4 +95,22 @@ public void removeLikeFromSpecificFilmByUser(Long filmId, Long userId) {
+
+    @Override
+    public List<FilmResponseDto> searchFilms(String query, String by) {
+        String[] searchBy = by.split(",");
```
выдели булеву переменную и в стриме выполнить все операциию Если не придумаешь, подскажу

rifnurV replied:
добавил вычисление в stream

rifnurV committed:
1. if(searchByTitle) в else добавил Collections.emptyList();
2. query оформил через """ sql query """
3. searchByTitle вычислил через stream
4. в постмане програл.

Тесты будут позже

rifnurV reviewed (commented):
внес изменения

rifnurV commented:
> Тесты + зелёная джоба из коллекции Постмана

тесты проходят кроме director. тесты напишу позже.

TatyanaEn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -117,4 +117,15 @@ public Set<Long> getFilmLikesByFilmId(Long filmId) {
+    public List<Film> searchFilms(String query, Boolean searchByTitle) {
+        List<Film> films = List.of();
+        if (searchByTitle){
+            films = jdbcTemplate.query("select * from film f where lower(f.name) like ? " +
```
Можно использовать регистронезависимый ilike вместо like.
Тогда отпадет необходимость принудительно приводить левую и правую часть в коде `lower(f.name) like ?` к нижнему регистру, 
то есть будет просто `f.name ilike ?`

rifnurV replied:
интересное решение, заменил

rifnurV committed:
заменил вместо like регистронезависимый ilike
добавил тест

rifnurV commented:
> Тесты + зелёная джоба по постман коллекции

добавил тест

TatyanaEn reviewed (approved):
замечаний нет

rifnurV committed:
Добавил тесты

rifnurV committed:
CheckStyle

rifnurV committed:
Дописал методы поиска
Добавил тест

rifnurV committed:
внес директора в схему

rifnurV committed:
Merge branch 'develop' of github.com:yusuf33nn/java-filmorate into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/DefaultFilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/api/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/api/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/in_memory/InMemoryFilmStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java

rifnurV committed:
Из FilmDbStorage убрал DirectorMapper

rifnurV committed:
directorMapper подкорректировал

rifnurV committed:
Merge branch 'develop' of github.com:yusuf33nn/java-filmorate into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/mapper/FilmMapper.java
#	src/main/java/ru/yandex/practicum/filmorate/model/dto/request/FilmRequestDto.java
#	src/main/java/ru/yandex/practicum/filmorate/service/DefaultFilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java
#	src/main/resources/schema.sql

rifnurV committed:
CodeStyle

rifnurV committed:
CheckStyle

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/service/DefaultDirectorService.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public DirectorResponseDto findDirectorById(Long id) {
+        if (id == null) {
+            return directorDbStorage.findAll().stream().map(directorMapper::toDto).findFirst().get();
```
вместо get() -> orElseThrow()

rifnurV replied:
поправил

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/service/DefaultDirectorService.java:
```diff
@@ -0,0 +1,79 @@
+        }
+        findDirectorById(directorId);
+        var updatedDirector = directorDbStorage.updateDirector(directorMapper.toEntity(director));
+        if (updatedDirector == null) {
```
log.error()

rifnurV replied:
удалил

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -79,6 +86,17 @@ public LinkedHashSet<Film> showMostPopularFilms(Integer count) {
+        Collection<Director> directorCollection = directorStorage.findDirectorsByParams(listDirectors);
+
+        if (listDirectors.size() != directorCollection.size()) {
+            throw new NotFoundException("Directors not found");
```
должен быть RuntimeException()

rifnurV replied:
поправил

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -100,12 +118,32 @@ public Film saveFilm(Film film) {
+        Collection<Director> directorCollection = directorStorage.findDirectorsByParams(listDirectors);
+
+        if (listDirectors.size() != directorCollection.size()) {
+            throw new NotFoundException("Directors not found");
```
тоже самое

rifnurV replied:
поправил

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -147,6 +199,77 @@ public Set<Long> getFilmLikesByFilmId(Long filmId) {
+
+        if (searchByTitle && searchByDirector) {
+            sql = """
+                    SELECT f.*
```
выделить в отдельные метод каждый блок If

rifnurV replied:
делить не нужно

yusuf33nn replied:
почему не нужно?

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/storage/db_storage/FilmDbStorage.java:
```diff
@@ -147,6 +199,77 @@ public Set<Long> getFilmLikesByFilmId(Long filmId) {
+        System.out.println("sortBy" + sortBy);
+        String sql;
+        if ("year".equals(sortBy)) {
+            sql = """
```
раздели на отдельные методы

rifnurV replied:
поправил

rifnurV reviewed (commented):
поправил

rifnurV committed:
Внес поправки

rifnurV committed:
CheckStyle

rifnurV committed:
Код вынес в отдельные методы

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -35,6 +36,22 @@ public ResponseEntity<List<FilmResponseDto>> showMostPopularFilms(int count) {
+    public ResponseEntity<List<FilmResponseDto>> searchFilmsByDirector(Long directorId, String sortBy) {
+        log.info("Getting films by director ID: {}, sorted by: {}", directorId, sortBy);
+
+        if (!"year".equals(sortBy) && !"likes".equals(sortBy)) {
```
хотели же вынести в сервис?

rifnurV replied:
перенес

rifnurV committed:
вернул log.error(errorMessage);

rifnurV committed:
проверку из контроллера вынес в сервис

rifnurV reviewed (commented):
внес поправки

yusuf33nn commented on src/main/java/ru/yandex/practicum/filmorate/service/DefaultFilmService.java:
```diff
@@ -97,4 +104,74 @@ public List<FilmResponseDto> findCommonFilms(Long userId, Long friendId) {
+                .map(String::toLowerCase)
+                .collect(Collectors.toSet());
+
+        if (!"year".equals(sortBy) && !"likes".equals(sortBy)) {
```
это должно быть в самом начале метода. Если проверка не проходит, то стримы выше были лишней операацией

rifnurV replied:
поправил

rifnurV committed:
проверку в сервисе DefaultFilmService метода searchFilmsByDirector поднял высше

rifnurV reviewed (commented):
поправил

rifnurV committed:
поправил метод public DirectorResponseDto findDirectorById(Long id)
